AdminServer/appscale/admin/service_manager.py

```diff
@@ -0,0 +1,409 @@
+            state.append(server)
+      except IOError as error:
+        if error.errno != errno.ENOENT:
+          raise
+
+    return state
+
+  def start(self):
+    """ Begin watching for assignments. """
+    logger.info('Starting ServiceManager')
+
+    # Ensure cgroup process containers exist.
+    for service_details in self.SERVICE_MAP.values():
+      cgroup_path = [CGROUP_DIR] + service_details['cgroup']
+      try:
+        os.makedirs(os.path.join(*cgroup_path))
+      except OSError as error:
+        if error.errno != errno.EEXIST:
+          raise
+
+    self.state = self.get_state(self._http_client)
+    self._zk_client.DataWatch(self._assignments_path,
+                              self._update_services_watch)
+    PeriodicCallback(self._groom_servers,
+                     self.GROOMING_INTERVAL * 1000).start()
+
+  @gen.coroutine
+  def _groom_servers(self):
+    """ Forgets about outdated servers and fulfills assignments. """
+    def outdated(server):
+      if (server.state == ServerStates.FAILED and
+          time.time() > server.failure_time + self.FAILED_SERVER_RETENTION):
+        return True
+
+      if server.state == ServerStates.STOPPED:
+        return True
+
+      return False
+
+    self.state = [server for server in self.state if not outdated(server)]
+    for service_type, options in self.assignments.items():
+      yield self._schedule_service(service_type, options)
+
+    for server in self.state:
+      if server.state != ServerStates.RUNNING:
+        continue
+
+      IOLoop.current().spawn_callback(server.ensure_running)
+
+  def _get_open_port(self):
+    """ Selects an available port for a server to use.
+
+    Returns:
+      An integer specifying a port.
+    """
+    assigned_ports = set(service.port for service in self.state)
+    port = self.START_PORT
+    while True:
+      # Skip ports that have been assigned.
+      if port in assigned_ports:
+        port += 1
+        continue
+
+      return port
+
+  @gen.coroutine
+  def _schedule_service(self, service_type, options):
+    """ Schedules servers to fulfill service assignment.
+
+    Args:
+      service_type: A string specifying the service type.
+      options: A dictionary specifying options to use when starting servers.
+    """
+    scheduled = [server for server in self.state
+                 if server.type == service_type and
+                 server.state in self.SCHEDULED_STATES]
+    to_start = options['count'] - len(scheduled)
+    if to_start < 0:
+      stopped = 0
+      for server in reversed(scheduled):
+        if stopped >= abs(to_start):
+          break
+
+        logger.info('Stopping {}'.format(server))
+        IOLoop.current().spawn_callback(server.stop)
+        stopped += 1
+
+      return
+
+    for _ in range(to_start):
+      port = self._get_open_port()
+      server_class = self.SERVICE_MAP[service_type]['server']
+      server = server_class(port, self._http_client, options['verbose'])
```

Comment:
Shouldn't service general options be stored outside assignments nodes which are node-specific, when `verbose` is service specific option (independent from nodes).

Reply:
I agree that the verbose flag should be stored separately from the assignment. It requires a bit of preparatory work on the logger object and setting up an additional watch in the datastore server. I'm inclined to leave this pull request with the instance options just so we can unblock datastore autoscaling, but I plan to follow up with an additional pull request to remove that option from the assignment.